src/GraphQL/Utilities/Federation/Proposal.cs

```diff
@@ -0,0 +1,243 @@
+        /// Public helper which will allow us to set "__AST_MetaField__" on a field which then will be used by FederatedSchemaPrinter
+        /// </summary>
+        /// <param name="type"></param>
+        /// <param name="name"></param>
+        /// <param name="value"></param>
+        public static void BuildAstMeta(this IProvideMetadata type, string name, string value = null)
+        {
+            // var definition = type.GetAstType<GraphQLObjectTypeDefinition>() ?? BuildGraphQLObjectTypeDefinition();
+            var definition = (GraphQLObjectTypeDefinition)type.GetMetadata<ASTNode>("__AST_MetaField__", () => BuildGraphQLObjectTypeDefinition());
+            var directive = BuildGraphQLDirective(name, value);
+            AddDirective(definition, directive);
+            //type.SetAstType(definition);
+            type.Metadata["__AST_MetaField__"] = definition;
+        }
+    }
+
+    /// <summary>
+    /// FieldBuillderExtensions, provide shortcuts for filling field metadata with expected objects, e.g. instead of:
+    /// Field().Name("foo").FieldType.Metadata["__AST_MetaField__"] = new GraphQLObjectTypeDefinition() {...};
+    /// we will have:
+    /// Field().Name("foo").Requires("bar");
+    /// </summary>
+    public static class FederatedFieldBuilderExtensions
+    {
+        private static FieldBuilder<TSourceType, TReturnType> BuildAstMeta<TSourceType, TReturnType>(FieldBuilder<TSourceType, TReturnType> fieldBuilder, string name, string value = null)
+        {
+            fieldBuilder.FieldType.BuildAstMeta(name, value);
+            return fieldBuilder;
+        }
+
+        public static FieldBuilder<TSourceType, TReturnType> Key<TSourceType, TReturnType>(this FieldBuilder<TSourceType, TReturnType> fieldBuilder, string fields) => BuildAstMeta(fieldBuilder, "key", fields);
+        public static FieldBuilder<TSourceType, TReturnType> Requires<TSourceType, TReturnType>(this FieldBuilder<TSourceType, TReturnType> fieldBuilder, string fields) => BuildAstMeta(fieldBuilder, "requires", fields);
+        public static FieldBuilder<TSourceType, TReturnType> Provides<TSourceType, TReturnType>(this FieldBuilder<TSourceType, TReturnType> fieldBuilder, string fields) => BuildAstMeta(fieldBuilder, "provides", fields);
+        public static FieldBuilder<TSourceType, TReturnType> External<TSourceType, TReturnType>(this FieldBuilder<TSourceType, TReturnType> fieldBuilder) => BuildAstMeta(fieldBuilder, "external");
+    }
+
+    /// <summary>
+    /// Can be replaced with ObjectGraphType extensions, the same as above - idea to allow us to define type directives, e.g.:
+    /// public class AcmeType: ObjectGraphType {
+    ///   public AcmeType() {
+    ///     // Metadata["__AST_MetaField__"] = new GraphQLObjectTypeDefinition() {...};
+    ///     Key("id");
+    ///   }
+    /// }
+    ///
+    /// TODO:
+    /// - there is only two cases: `type Account @key(fields:"id")` and `extend type Account @key(fields:"id")`
```

Comment:
hey @mac2000 
can you give me a general idea on how I can do the `extend type Account {}` in code first? I tried a bunch of things and couldn't succeed.